website/docs: add docs for source switch expression policy (#16878)
goauthentik/authentik: website/docs/customize/policies/expression/source_switch.md

---
title: Switch which source is used based on email address
---

> You can use an expression policy to determine with [source](../../../../users-sources/sources/) (a set of user credentials and data, stored in authentik, Google, GitHub, etc) is used for a particular user, based on which email address the user enters when they log in and authenticate (using the authn flow).

Reviewer: I suggest removing "a set of user credentials and data, stored in" because we're already linking to the source document. Keeping it in makes the sentence hard to read IMO. But removing it isn't essential.

```suggestion
You can use an expression policy to determine which [source](../../../../users-sources/sources/) (authentik, Google, GitHub, etc) is used for a particular user, based on the domain of their email address.
```

> To define which source is used for a specific user, [create an expression policy](../working_with_policies.md#create-a-policy) that defines the possible options for the source, and the logic to determine the domain based on the email address and then "switch" the user to the desired source.

Reviewer: ```suggestion
To switch which source is used for a specific user based on their email domain, [create an expression policy](../working_with_policies.md#create-a-policy) that:

    1. Maps the desired source for each user domain.
    2. Determines the user's domain based on their email address.
    3. Then "switches" the user to the desired source.
```

## Example expression

```python
# This is a mapping of domains to sources
# the key is a domain for the user and the value is the `slug` of the source to redirect to
source_email_map = {
    "foo.bar.com": "entra-foo",
    "bar.baz.com": "entra-bar",
}

user_email = request.context["pending_user_identifier"]

_username, _, domain = user_email.partition("@")
source = source_email_map.get(domain)
if not source:
    return True
plan = request.context.get("flow_plan")
if not plan:
    return False
# For OIDC
# plan.redirect(f"/source/oauth/login/{source}/")
# For SAML
plan.redirect(f"/source/saml/{source}")
return False

> ```

Reviewer: We'll need to add a section outlining where to bind the policy
